File: examples/export-pages-and-assets/frameworks/playwright/csharp/DownloadImages.cs
// Exports a page's HTML and linked assets (CSS, JS, images) to a local directory
// by intercepting network responses as the page loads.
//
// Install: dotnet add package Microsoft.Playwright && playwright install chromium
// Run:     dotnet run

using Microsoft.Playwright;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

class DownloadImages
{
    static async Task Main()
    {
        const string TOKEN = "YOUR_API_TOKEN_HERE";
        string WS_ENDPOINT = $"wss://production-sfo.browserless.io?token={TOKEN}";

        using var playwright = await Playwright.CreateAsync();
        var browser = await playwright.Chromium.ConnectOverCDPAsync(WS_ENDPOINT);

        try
        {
            var context = browser.Contexts[0];
            var page = await context.NewPageAsync();
            var assetTasks = new List<Task<(string url, byte[] buf)?>>();

            page.Response += (_, response) =>
            {
                var type = response.Request.ResourceType;
                if (type is "stylesheet" or "script" or "image" or "font")
                {
                    var url = response.Url;
                    assetTasks.Add(response.BodyAsync()
                        .ContinueWith(t => t.IsCompletedSuccessfully
                            ? ((string, byte[])?) (url, t.Result)
                            : null));
                }
            };

            await page.GotoAsync("https://example.com",
                new() { WaitUntil = WaitUntilState.NetworkIdle });

            var html = await page.ContentAsync();
            var assets = (await Task.WhenAll(assetTasks)).Where(a => a.HasValue).Select(a => a!.Value);

            Directory.CreateDirectory("page");
            await File.WriteAllTextAsync("page/index.html", html);
            Console.WriteLine("Saved page/index.html");

            int i = 0;
            foreach (var (url, buf) in assets)
            {
                string ext = Path.GetExtension(new Uri(url).AbsolutePath);
                string filename = $"page/asset-{i}{ext}";
                await File.WriteAllBytesAsync(filename, buf);
                Console.WriteLine($"Saved {filename}");
                i++;
            }
        }
        finally
        {
            await browser.CloseAsync();
        }
    }
}

File: examples/export-pages-and-assets/frameworks/playwright/java/DownloadImages.java
// Exports a page's HTML and linked assets (CSS, JS, images) to a local directory
// by intercepting network responses as the page loads.
//
// Add to pom.xml:
//   <dependency>
//     <groupId>com.microsoft.playwright</groupId>
//     <artifactId>playwright</artifactId>
//     <version>1.44.0</version>
//   </dependency>
//
// Compile: mvn compile
// Run:     mvn exec:java

import com.microsoft.playwright.*;
import java.nio.file.*;
import java.net.URI;
import java.util.ArrayList;
import java.util.List;

public class DownloadImages {
    record Asset(String url, byte[] buf) {}

    public static void main(String[] args) throws Exception {
        String TOKEN = "YOUR_API_TOKEN_HERE";
        String WS_ENDPOINT = "wss://production-sfo.browserless.io?token=" + TOKEN;

        try (Playwright playwright = Playwright.create()) {
            Browser browser = playwright.chromium().connectOverCDP(WS_ENDPOINT);
            try {
                BrowserContext context = browser.contexts().get(0);
                Page page = context.newPage();
                List<Asset> assets = new ArrayList<>();

                page.onResponse(response -> {
                    String type = response.request().resourceType();
                    if (List.of("stylesheet", "script", "image", "font").contains(type)) {
                        try {
                            assets.add(new Asset(response.url(), response.body()));
                        } catch (Exception ignored) {}
                    }
                });

                page.navigate("https://example.com");
                page.waitForLoadState(LoadState.NETWORKIDLE);

                String html = page.content();

                Files.createDirectories(Path.of("page"));
                Files.writeString(Path.of("page/index.html"), html);
                System.out.println("Saved page/index.html");

                for (int i = 0; i < assets.size(); i++) {
                    String urlPath = new URI(assets.get(i).url()).getPath();
                    int dot = urlPath.lastIndexOf('.');
                    String ext = dot >= 0 ? urlPath.substring(dot) : "";
                    Path dest = Path.of("page/asset-" + i + ext);
                    Files.write(dest, assets.get(i).buf());
                    System.out.println("Saved " + dest);
                }
            } finally {
                browser.close();
            }
        }
    }
}
